renderdoc/driver/metal/metal_core.cpp

```diff
@@ -398,12 +398,49 @@ void WrappedMTLDevice::CaptureCmdBufSubmit(MetalResourceRecord *record)
       MetalResourceRecord *refRecord = GetResourceManager()->GetResourceRecord(id);
       if(refRecord->m_Type == eResBuffer)
       {
-        // TODO: capture CPU modified buffers
+        MetalBufferInfo *bufInfo = refRecord->bufInfo;
+        if(bufInfo->storageMode == MTL::StorageModeShared)
+        {
+          size_t diffStart = 0;
+          size_t diffEnd = 0;
+          bool foundDifference = true;
+          if(bufInfo->baseSnapshot.isEmpty())
+          {
+            bufInfo->baseSnapshot.assign(bufInfo->data, bufInfo->length);
+            diffEnd = bufInfo->length;
```

Comment:
There is a small but observable in practice race condition here, because what we're serialising is not exactly what we snapshot here for further comparisons.

Any data the application wants to be available to the GPU has to be modified before this point and any modifications during/after this point are only available later, so we have some freedom in what we choose to save. If we were to snapshot, then serialise the current buffer contents, it's possible that another thread is modifying it in between and so what is serialised in this first modify chunk is not what gets compared against next time.

Because you're snapshotting before serialising, this is less serious but I think could still cause problems in terms of an A-B-A issue. If the buffer contains 1234 and is snapshotted like that, then there's a write between the snapshot and serialise which changes it to 4321, we serialise 4321, then before the next sync/compare point it's written back to 1234 we'd not be able to detect that change and would potentially 'leave' it wrongly as 4321. This is not necessarily invalid on the application's part if that part of the buffer is not used by the GPU until the appropriate time. The other way around is much worse because it doesn't require an A-B-A just a late write between serialise and snapshot.

The approach taken in the other drivers is to copy the first comparison snapshot out of what we serialised so we always compare against what we last serialised, even though it's still somewhat arbitrary in the face of ongoing parallel writes, and assume the application did things correctly and any such writes are not needed for this submission and are fine to pick up in the next one.

I also notice that baseSnapshot is not updated after the first write, which will also cause similar kinds of problems. The snapshot needs to be updated each time so we compare against what was there last to determine what is new/changed and needs to be serialised.